# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/renderer/shade.rs
# Commit: 76e2b9f2481d
# Full Hash: 76e2b9f2481dbcacbc2e651a1767703edb37992e
# Author: Jamie Nicol <jnicol@mozilla.com>
# Date: 2023-12-19 05:06:09
# Regressor Bug: 1866020
# File Overlap Count: 1
# Description:
#   Bug 1870413 - Allow creating TEXTURE_EXTERNAL_BT709 shaders on devices without image_external_essl3. r=gfx-reviewers,lsalzman
#   
#   An oversight when updating the allowed shaders list in bug 1866020
#   assumed that devices without GL_OES_EGL_image_external_essl3 would
#   also not have GL_EXT_YUV_target. This turns out not to always be the
# ==============================================================================

diff -r c909d207f316 -r 76e2b9f2481d gfx/wr/webrender/src/renderer/shade.rs
--- a/gfx/wr/webrender/src/renderer/shade.rs	Mon Dec 18 17:56:42 2023 +0000
+++ b/gfx/wr/webrender/src/renderer/shade.rs	Mon Dec 18 18:04:56 2023 +0000
@@ -661,7 +661,7 @@
         } else {
             TextureExternalVersion::ESSL1
         };
-        let mut shader_flags = get_shader_feature_flags(gl_type, texture_external_version);
+        let mut shader_flags = get_shader_feature_flags(gl_type, texture_external_version, device);
         shader_flags.set(ShaderFeatureFlags::ADVANCED_BLEND_EQUATION, use_advanced_blend_equation);
         shader_flags.set(ShaderFeatureFlags::DUAL_SOURCE_BLENDING, use_dual_source_blending);
         shader_flags.set(ShaderFeatureFlags::DITHERING, options.enable_dithering);
@@ -1355,7 +1355,7 @@
             TextureExternalVersion::ESSL1
         };
 
-        let feature_flags = get_shader_feature_flags(gl_type, texture_external_version);
+        let feature_flags = get_shader_feature_flags(gl_type, texture_external_version, device);
         let shader_list = get_shader_features(feature_flags);
 
         for _ in 0..IMAGE_BUFFER_KINDS.len() {
@@ -1485,18 +1485,23 @@
     }
 }
 
-fn get_shader_feature_flags(gl_type: GlType, texture_external_version: TextureExternalVersion) -> ShaderFeatureFlags {
+fn get_shader_feature_flags(
+    gl_type: GlType,
+    texture_external_version: TextureExternalVersion,
+    device: &Device
+) -> ShaderFeatureFlags {
     match gl_type {
         GlType::Gl => ShaderFeatureFlags::GL,
         GlType::Gles => {
-            let texture_external_flag = match texture_external_version {
-                TextureExternalVersion::ESSL3 => {
-                    ShaderFeatureFlags::TEXTURE_EXTERNAL
-                        | ShaderFeatureFlags::TEXTURE_EXTERNAL_BT709
-                }
+            let mut flags = ShaderFeatureFlags::GLES;
+            flags |= match texture_external_version {
+                TextureExternalVersion::ESSL3 => ShaderFeatureFlags::TEXTURE_EXTERNAL,
                 TextureExternalVersion::ESSL1 => ShaderFeatureFlags::TEXTURE_EXTERNAL_ESSL1,
             };
-            ShaderFeatureFlags::GLES | texture_external_flag
+            if device.supports_extension("GL_EXT_YUV_target") {
+                flags |= ShaderFeatureFlags::TEXTURE_EXTERNAL_BT709;
+            }
+            flags
         }
     }
 }
